Add Environment pipeline representation

violalyu opened the pull request:
**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

tvalentyn committed:
Clean up a few places where we pass arbitrary kwargs to PTransform constructor.

robertwb committed:
Implement ZipFile backed artifact service.

robertwb committed:
Make compatible with Java artifact services.

Rather than deriving the path from the retrieval token directly,
store a ProxyManifest that contains the artifact name to uri
mappings.

robertwb committed:
[BEAM-8372] Job server submitting UberJars directly to Flink Runner.

One can now use the FlinkRunner with the --flink_master_url=flink-master-url
and an UberJar will automatically be created with the artifacts and pipeline
definition, uploaded to the master, and started.

Commit:
[BEAM-8390] usable useCorrelationId with RabbitMq

robertwb committed:
Account for post BEAM-8183 file structure.

violalyu committed:
[BEAM-8402] Create a class hierarchy to represent environments

As a first step towards making it possible to assign different environments to sections of a pipeline, we first need to expose environment classes to the pipeline API. Unlike PTransforms, PCollections, Coders, and Windowings, environments exists solely in the portability framework as protobuf objects. By creating a hierarchy of "native" classes that represent the various environment types – external, docker, process, etc – users will be able to instantiate these and assign them to parts of the pipeline.

violalyu commented:
@chadrik Updated!

chadrik commented on sdks/python/apache_beam/runners/portability/fn_api_runner.py:
```diff
@@ -334,7 +335,7 @@ def __init__(
     self._default_environment = (
         default_environment
-        or beam_runner_api_pb2.Environment(urn=python_urns.EMBEDDED_PYTHON))
+        or environments.Environment(urn=python_urns.EMBEDDED_PYTHON))
```
missed this one

chadrik commented on sdks/python/apache_beam/transforms/environments.py:
```diff
@@ -0,0 +1,276 @@
+
+
+class PortableEnvironment(Environment):
+
```
I don't understand the purpose of this class.  The 'portable' urn did not exist before, so I think we may be doing something wrong by inventing it.  Why not use `ExternalEnvironment`?

violalyu replied:
`ExternalEnvironment` requires `ExternalPayload` type, and in the `PortableRunner._create_environment` it was implying that if the `portable_options.environment_type` was passed in a urn that starts with `beam:env:` but not one of the urns in `DOCKER`, `PROCESS`, `EXTERNAL`, then it creates just an environment with the given urn and config as payload.

chadrik replied:
my impression is that anything starting with `beam:env:` should be in python_urns.  let's go with that assumption.

chadrik commented on sdks/python/apache_beam/runners/portability/portable_runner.py:
```diff
@@ -95,65 +96,28 @@ def default_docker_image():
+    else:
+      try:
+        environment_urn = environments.Environment.get_urn_from_type(
+          environment_type)
```
I'd like to get rid of this lookup.  Instead, I'd like to do something like:

```python
environment_type = portable_options.environment_type
if not environment_type:
  environment_urn = common_urns.environments.DOCKER.urn
elif environment_type.startswith('beam:env:'):
  environment_urn = environment_type
else:
  # e.g. handle LOOPBACK -> EXTERNAL
  environment_type = ENV_TYPE_ALIASES.get(environment_type, environment_type)
  environment_urn = 'beam:env:' + environment_type.lower()
```

this will allow us to completely get rid of the `Environment.env_type` class attribute.

chadrik replied:
and also we can get rid of `get_urn_from_type()`

chadrik replied:
and `_type_to_urn`

chadrik commented on sdks/python/apache_beam/transforms/environments.py:
```diff
@@ -0,0 +1,276 @@
+      return register
+
+  @classmethod
+  def register_type_and_urn(cls, env_cls):
```
call this `register_environment`

11moon11 committed:
Replaced depricated calls

rahul8383 committed:
[BEAM-8394]: Add withDataSourceConfiguration() method in JdbcIO.ReadRows

mxm committed:
[BEAM-8403] Guard request id generation to prevent concurrent worker access

There is a race condition in GrpcStateRequestHandler which surfaced after the
recent changes to process append/clear state request asynchronously. The race
condition can occur if multiple Runner workers process a transform with state
requests with the same SDK Harness. The SDK Harness has multiple workers threads
which can access the id generation code in parallel.

For example, this setup occurs with Flink when a TaskManager has multiple task
slots and two or more of those slots process the same stateful stage against an
SDK Harness.

robertwb committed:
Merge pull request #9775 from robertwb/easy-flink

[BEAM-8372] Job server submitting UberJars directly to Flink Runner.

aaltay committed:
Fixing a logging bug (#9795)

aaltay committed:
Fixing a logging bug in auth (#9796)

apilloud committed:
Merge pull request #9798: [BEAM-8404] Replaced deprecated calls

chadrik commented on sdks/python/apache_beam/transforms/environments.py:
```diff
@@ -0,0 +1,258 @@
+
+@Environment.register_environment
+class PythonSDKEnvironment(Environment):
+
```
add `urn = None` here and raise a `ValueError` if `self.urn is None` in `to_runner_api_parameter`

chadrik replied:
forget the first part, it's already on `Environment`.

chadrik commented on sdks/python/apache_beam/transforms/environments.py:
```diff
@@ -0,0 +1,258 @@
+
+  @classmethod
+  def register_environment(cls, env_cls):
+    cls._urn_to_env_cls[env_cls.urn] = env_cls
```
raise `ValueError` if `env_cls.urn is Non`

chadrik commented on sdks/python/apache_beam/transforms/environments.py:
```diff
@@ -0,0 +1,258 @@
+  @classmethod
+  def from_runner_api(cls, proto, context):
+    if proto is None or not proto.urn:
+      return None
```
is this a pattern that's used elsewhere?  not sure what the circumstances would be when either of these would be allowed to be `None`

violalyu replied:
It is the pattern from `PTransform.from_runner_api`

chadrik commented on sdks/python/apache_beam/transforms/environments.py:
```diff
@@ -0,0 +1,258 @@
+      raise
+
+  @classmethod
+  def from_options(cls, options):
```
docstring

chadrik commented on sdks/python/apache_beam/transforms/environments.py:
```diff
@@ -0,0 +1,258 @@
+PORTABLE_ENVIRONMENT = 'portable'
+
+
+class Environment(object):
```
docstring

chadrik commented on sdks/python/apache_beam/transforms/environments.py:
```diff
@@ -0,0 +1,258 @@
+  _known_urns = {}
+  _urn_to_env_cls = {}
+
+  urn = None
```
add a comment explaining that this must be provided by subclasses

chadrik commented on sdks/python/apache_beam/transforms/environments.py:
```diff
@@ -0,0 +1,258 @@
+  @Environment.register_urn(common_urns.environments.DOCKER.urn,
+                            beam_runner_api_pb2.DockerPayload)
+  def from_runner_api_parameter(payload, components, context):
+    return DockerEnvironment(container_image=payload.container_image)
```
there's one problem with this design that still bothers me a little:  the user has to provide the urn in 3 places. 

looking around the code, I see it's pretty common for users to provide the urn at least twice:

```python
  def to_runner_api_parameter(self, context):
    return (common_urns.fixed_windows.urn,
            standard_window_fns_pb2.FixedWindowsPayload(
                size=proto_utils.from_micros(
                    duration_pb2.Duration, self.size.micros),
                offset=proto_utils.from_micros(
                    timestamp_pb2.Timestamp, self.offset.micros)))

  @urns.RunnerApiFn.register_urn(
      common_urns.fixed_windows.urn,
      standard_window_fns_pb2.FixedWindowsPayload)
  def from_runner_api_parameter(fn_parameter, unused_context):
    return FixedWindows(
        size=Duration(micros=fn_parameter.size.ToMicroseconds()),
        offset=Timestamp(micros=fn_parameter.offset.ToMicroseconds()))
```

Environments are different because we have this need to also find the class based on urn.   This makes them more like `Coders`, which have a shortcut for creating the `to_runner_api_parameter` / `from_runner_api_parameter` pair:

```python
class StrUtf8Coder(Coder):
  """A coder used for reading and writing strings as UTF-8."""

  def encode(self, value):
    return value.encode('utf-8')

  def decode(self, value):
    return value.decode('utf-8')

  def is_deterministic(self):
    return True

  def to_type_hint(self):
    return unicode

Coder.register_structured_urn(
    common_urns.coders.STRING_UTF8.urn, StrUtf8Coder)
```

```python
  @staticmethod
  def register_structured_urn(urn, cls):
    # type: (str, Type[Coder]) -> None
    """Register a coder that's completely defined by its urn and its
    component(s), if any, which are passed to construct the instance.
    """
    def to_runner_api_parameter(self, context):
      # type: (Coder, PipelineContext) -> Tuple[str, Any, Sequence[Coder]]
      return (urn, None, self._get_component_coders())

    setattr(cls, 'to_runner_api_parameter', to_runner_api_parameter)

    # pylint: disable=unused-variable
    @Coder.register_urn(urn, None)
    def from_runner_api_parameter(unused_payload, components, unused_context):
      if components:
        return cls(*components)
      else:
        return cls()
```

Let's think a bit more about how we can simplify this while also keeping with the prevailing patterns of the beam code.

violalyu replied:
In this [commit](28c66f78c4d61c6c6cc68dd3c2b97ae15617b2f6) I've attempted to use such pattern for `PythonSDKEnvironment` subclasses, since they fit the pattern most; For other Environments, one thing is that the payload and their constructor signature is different, which makes it harder to fit in

pabloem committed:
Showing the Java environment used by Dataflow in the UI (#8680)

* Showing the Java environment used by Dataflow in the UI via userAgent

violalyu commented on sdks/python/apache_beam/transforms/environments.py:
```diff
@@ -76,7 +98,8 @@ def to_runner_api(self, context):
         if isinstance(typed_param, message.Message)
-        else typed_param.encode('utf-8') if isinstance(typed_param, str)
+        else typed_param.encode('utf-8') if isinstance(typed_param,
+                                                       (str, type(u'')))
```
@chadrik I'm not sure what would be the canonical way of doing this? since python3 doesn't have unicode / basestring anymore, but we do want python2 unicode type string to be encoded?

chadrik replied:
It's the responsibility of `to_runner_api_parameter` to return the same type it was registered with, in this case `bytes`

```python
@Environment.register_python_environment(python_urns.EMBEDDED_PYTHON, bytes)
```

additionally, if we're following the precedent set by other registration decorators, the only valid types that can be passed are message.Message, bytes, or None. 

I saw that `PTransform.to_runner_api` has the pattern that you're using, but I don't understand it.  To me it seems like it would only work on python3 (where `str` is actually `unicode`).   Plus, I don't see anywhere that a unicode payload is used.

Another thing to note about `PTransform.register_urn` is that it implements a special pattern that works for classes or methods:

```python

  @classmethod
  def register_urn(cls, urn, parameter_type, constructor=None):
    def register(constructor):
      if isinstance(constructor, type):
        constructor.from_runner_api_parameter = register(
            constructor.from_runner_api_parameter)
        # pylint isn't smart enough to recognize when this is used
        # on a class or a method, and will emit a no-self-warning
        # in the latter case.  Rather than suppressing this at each
        # use, we fool it here through some dynamic patching that
        # pylint will also not understand.

        @contextlib.contextmanager
        def fake_static_method():
          actual_static_method = staticmethod
          globals()['staticmethod'] = lambda x: x
          yield
          globals()['staticmethod'] = actual_static_method
        with fake_static_method():
          return staticmethod(constructor)
      else:
        cls._known_urns[urn] = parameter_type, constructor
        return staticmethod(constructor)
    if constructor:
      # Used as a statement.
      register(constructor)
    else:
      # Used as a decorator.
      return register
```

That could be a good fit, allowing us to merge `register_environment` into `register_urn`

violalyu replied:
Updated in https://github.com/violalyu/beam/pull/1/commits/7e32a4ccdc41d176fc349537d5da31de05129277

robertwb committed:
Merge pull request #9674 from tvalentyn/ptransform_cleanup

Clean up a few places where we pass arbitrary kwargs to PTransform constructor.

iemejia committed:
Merge pull request #9799: [BEAM-8394]: Add withDataSourceConfiguration() method in JdbcIO.ReadRows

mxm committed:
Merge pull request #9800: [BEAM-8403] Guard request id generation to prevent concurrent worker access

iemejia committed:
Merge pull request #9782: [BEAM-8390] Allow specifying useCorrelationId for RabbitMqIO.Read

lostluck committed:
[Go SDK] Fix subtle split error. Add more tests. (#9750)

* [Go SDK] Fix subtle split error. Add more tests.

There was an error in how we handle splitting from datasources, where we were permitting splitting on the current in progress element, which is against the definition of the residual. The first residual is the first unprocessed element. This happened rarely, and was thus hard to detect.

I figured out a way to block processing so we could send "in process" split requests and validate that we never split on the current element. This also lets us validate the Progress count (that is, the number of completely processed elements) is accurate, since we know which index on which we've blocked, and how many came before.

mwalenia committed:
[BEAM-8385] Add a KafkaIOIT hash for 100M records (#9770)

* [BEAM-8385] Tweak Kafka test cluster configuration

The default replication factor of the test topic changed to 1 in order to overcome synchronisation issues. Changed log retention settings to ensure that no records would be purged during the test.

* [BEAM-8385] Add PCollection hash for 100 000 000 records in the IOIT.

chadrik committed:
[BEAM-8350] Upgrade to Pylint 2.4 (#9725)

chadrik commented on sdks/python/apache_beam/transforms/environments.py:
```diff
@@ -0,0 +1,331 @@
+    return cls(container_image=options.environment_config)
+
+
+Environment.register_urn(common_urns.environments.DOCKER.urn,
```
I believe we should still be able to use this as a decorator, as with `PTransform.register_urn`

chadrik commented on sdks/python/apache_beam/transforms/environments.py:
```diff
@@ -0,0 +1,331 @@
+
+@Environment.register_python_environment(python_urns.SUBPROCESS_SDK, bytes)
+class SubprocessSDKEnvironment(PythonSDKEnvironment):
+  pass
```
I have an idea to further simplify this:

```python

@Environment.register_urn(python_urns.EMBEDDED_PYTHON, None)
class EmbeddedPythonEnvironment(Environment):
  def to_runner_api_parameter(self, context):
    return (common_urns.environments.EMBEDDED_PYTHON.urn, None)

  @staticmethod
  def from_runner_api_parameter(unused_payload, context):
    # parse payload into args
    return EmbeddedPythonEnvironment()


@Environment.register_python_environment(python_urns.EMBEDDED_PYTHON_GRPC,
                                         bytes)
class EmbeddedPythonGrpcEnvironment(Environment):
  def __init__(num_workers=None, state_cache_size=None):
      self.num_workers = num_workers
      self.state_cache_size = state_cache_size

  def to_runner_api_parameter(self, context):
    if self.num_workers and self.state_cache_size:
      payload = b'%d%d' % (self.num_workers, self.state_cache_size)
    elif:
      raise ???
    else:
      payload = b''

    return (common_urns.environments.EMBEDDED_PYTHON_GRPC.urn, payload)

  @staticmethod
  def from_runner_api_parameter(payload, context):
    # parse payload into args
    return EmbeddedPythonGrpcEnvironment(num_workers=x, state_cache_size=y)


@Environment.register_python_environment(python_urns.SUBPROCESS_SDK, bytes)
class SubprocessSDKEnvironment(Environment):
  def __init__(self, command_string):
    self.command_string = command_string

  def to_runner_api_parameter(self, context):
    return (common_urns.environments.SUBPROCESS_SDK.urn, 
                 self.command_string if isinstance(self.command_string, bytes)
                 else self.command_string.encode('utf-8'))

  @staticmethod
  def from_runner_api_parameter(payload, context):
    # parse payload into args
    return SubprocessSDKEnvironment(payload.decode('utf-8'))
```

violalyu replied:
Updated in https://github.com/violalyu/beam/pull/1/commits/45e955a6de7c944a1b540987d59217ab627ddab9

pabloem committed:
[BEAM-8367] Using insertId for BQ streaming inserts (#9797)

* Using insertId for BQ streaming inserts

* Adding unittest for BQ Insert Ids

chadrik commented on sdks/python/apache_beam/transforms/environments.py:
```diff
@@ -0,0 +1,373 @@
+      num_workers, state_cache_size = payload.decode('utf-8').split(',')
+      return EmbeddedPythonGrpcEnvironment(num_workers=int(num_workers),
+                                           state_cache_size=int(
+                                               state_cache_size))
```
kind of a strange line break.   break on the first paren.

chadrik commented on sdks/python/apache_beam/transforms/environments.py:
```diff
@@ -0,0 +1,373 @@
+      # We want to make sure that the environment stays the same through the
+      # roundtrip to runner api, so here we don't want to set default for the
+      # other if only one of num workers or state cache size is set
+      raise ValueError('Must provide worker num and state cache size.')
```
how about this:

```python
    if self.num_workers is None and self.state_cache_size is None:
      payload = b''
    elif self.num_workers is not None and self.state_cache_size is not None:
      payload = b'%d,%d' % (self.num_workers, self.state_cache_size)
    else:
      # We want to make sure that the environment stays the same through the
      # roundtrip to runner api, so here we don't want to set default for the
      # other if only one of num workers or state cache size is set
      raise ValueError('Must provide worker num and state cache size.')
```

chadrik commented on sdks/python/apache_beam/transforms/environments_test.py:
```diff
@@ -0,0 +1,67 @@
+        EmbeddedPythonEnvironment(),
+        EmbeddedPythonGrpcEnvironment(),
+        EmbeddedPythonGrpcEnvironment(num_workers=2, state_cache_size=0),
+        SubprocessSDKEnvironment(command_string='foo')):
```
let's use a unicode string here to be sure:

```python
command_string=u'foö'
```

chadrik commented:
looks great.  fix those last issues and run the lint job.

```
tox -e py27-lint
```

or this: if you're working from the very latest master:

```
tox -e py37-lint
```
